trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/eldarica/EldaricaChcScript.java

```diff
@@ -0,0 +1,370 @@
+import java.util.stream.IntStream;
+
+import ap.SimpleAPI;
+import ap.parser.IAtom;
+import ap.parser.IFormula;
+import ap.terfor.preds.Predicate;
+import de.uni_freiburg.informatik.ultimate.core.lib.exceptions.ToolchainCanceledException;
+import de.uni_freiburg.informatik.ultimate.core.model.services.ILogger;
+import de.uni_freiburg.informatik.ultimate.core.model.services.IUltimateServiceProvider;
+import de.uni_freiburg.informatik.ultimate.lib.chc.Derivation;
+import de.uni_freiburg.informatik.ultimate.lib.chc.HcPredicateSymbol;
+import de.uni_freiburg.informatik.ultimate.lib.chc.HcSymbolTable;
+import de.uni_freiburg.informatik.ultimate.lib.chc.HornClause;
+import de.uni_freiburg.informatik.ultimate.lib.chc.IChcScript;
+import de.uni_freiburg.informatik.ultimate.lib.chc.eldarica.Backtranslator.IBoundVariableContext;
+import de.uni_freiburg.informatik.ultimate.logic.Model;
+import de.uni_freiburg.informatik.ultimate.logic.Script;
+import de.uni_freiburg.informatik.ultimate.logic.Script.LBool;
+import de.uni_freiburg.informatik.ultimate.logic.Term;
+import de.uni_freiburg.informatik.ultimate.logic.TermVariable;
+import de.uni_freiburg.informatik.ultimate.smtsolver.external.FunctionDefinition;
+import de.uni_freiburg.informatik.ultimate.smtsolver.external.ModelDescription;
+import de.uni_freiburg.informatik.ultimate.util.Lazy;
+import lazabs.GlobalParameters;
+import lazabs.horn.bottomup.HornClauses.Clause;
+import lazabs.horn.bottomup.SimpleWrapper;
+import lazabs.prover.Tree;
+import scala.Option;
+import scala.Tuple2;
+import scala.collection.JavaConverters;
+import scala.collection.Seq;
+import scala.collection.immutable.List;
+
+/**
+ * Provides access to the eldarica constraint Horn solver.
+ *
+ * https://github.com/uuverifiers/eldarica/
+ *
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ */
+public class EldaricaChcScript implements IChcScript, AutoCloseable {
+	private final IUltimateServiceProvider mServices;
+	private final ILogger mLogger;
+	private final Script mScript;
+	private final SimpleAPI mPrincess;
+
+	private final long mDefaultQueryTimeout;
+	private boolean mProduceModels = false;
+	private boolean mProduceDerivations = false;
+	private boolean mProduceUnsatCores = false;
+
+	private Translator mTranslator;
+	private Map<Clause, HornClause> mClauseMap;
+
+	private LBool mLastResult = null;
+	private Lazy<ModelDescription> mLastModel;
+	private Lazy<Derivation> mLastDerivation;
+	private Lazy<Set<HornClause>> mLastUnsatCore;
+
+	public EldaricaChcScript(final IUltimateServiceProvider services, final Script script) {
+		this(services, script, -1L);
+	}
+
+	public EldaricaChcScript(final IUltimateServiceProvider services, final Script script,
+			final long defaultQueryTimeout) {
+		mServices = services;
+		mLogger = services.getLoggingService().getLogger(getClass());
+		mScript = script;
+		mPrincess = SimpleAPI.apply(SimpleAPI.apply$default$1(), SimpleAPI.apply$default$2(),
```

Comment:
what is happening here?

Reply:
An instance of the `SimpleAPI`-class is created, with default parameters (unfortunately, scala's default parameter values are somewhat verbose to use in Java). The class belongs to the API of eldarica's underlying SMT solver princess, which is used to create terms and formulae.